### PR TITLE
Implement filtering for /roomId/messages

### PR DESCRIPTION
**Pull request by hulthe**
Make [`GET /_matrix/client/r0/rooms/{roomId}/messages`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-rooms-roomid-messages) accept a `filter` parameter in
accordance with the spec (hopefully).

Related issue: #587 

This is my first PR to a matrix-project, so there are a couple of thing I'm unsure about:

- [`Database.GetEventsInTopologicalRange`][topological] doesn't support filters, so if the client specifies a
  topological `from`-token filtering wont work. Not sure if this is intended.

- `limit` can be set both as a query parameter, and as a part of RoomEventFilter.
  The spec doesn't seem to specify which takes precedence if both are specified.
  For now I've assumed that the query parameter takes precedence.

- Curiously, the default `limit` value in [`gomatrixserverlib.RoomEventFilter`][REF] is 20. I'm not sure
  if this is something I should adhere to, or event if it conforms to the spec, since the default for
  the query parameter is 10.

- Unlike [`/sync`][sync], the spec doesn't seem to allow for the filter payload to be a `filter_id`
  instead of a JSON RoomEventFilter. Is this intended, or a spec oversight?

[topological]: https://github.com/matrix-org/dendrite/blob/6e44450cc9b541e5e79b7e601462fd3d3cf35fe5/syncapi/storage/interface.go#L110 "GetEventsInTopologicalRange"
[REF]: https://github.com/matrix-org/gomatrixserverlib/blob/18abed1923fd09d5302090c8e91fab384b6cefb9/filter.go#L141 "RoomEventFilter"
[sync]: https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-sync "/sync spec"

### SyTest
Sadly, no new tests are passing (there doesn't seem to be many tests which handle filtering /messages),
although a couple of tests seem to be closer to passing than before:
- `Ephemeral messages received from clients are correctly expired`
- `Ephemeral messages received from servers are correctly expired`

These now fail for other reasons than filtering not working :)

### Pull Request Checklist

* [x] I have added any new tests that need to pass to `sytest-whitelist` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/docs/CONTRIBUTING.md#sign-off)

Signed-off-by: `Joakim Hulthe <joakim@hulthe.net>`

### CODE REVIEW

**Comment by hulthe** [comment]
Not sure why SyTest is failing in the CI. It seems to work on my machine :confused:

File: syncapi/routing/messages.go [hunk 1]
```diff
@@ -234,19 +249,17 @@ func (r *messagesReq) retrieveEvents() (
...
 	if r.fromStream != nil {
 		toStream := r.to.StreamToken()
 		streamEvents, err = r.db.GetEventsInStreamingRange(
-			r.ctx, r.fromStream, &toStream, r.roomID, &eventFilter, r.backwardOrdering,
+			r.ctx, r.fromStream, &toStream, r.roomID, &r.filter, r.backwardOrdering,
 		)
 	} else {
 		streamEvents, err = r.db.GetEventsInTopologicalRange(
-			r.ctx, r.from, r.to, r.roomID, r.limit, r.backwardOrdering,
+			r.ctx, r.from, r.to, r.roomID, r.filter.Limit, r.backwardOrdering,
```
Comment (Kegsay) [inline syncapi/routing/messages.go:262]: We need the filter object passed through to `GetEventsInTopologicalRange` and for it to be implemented there unfortunately!

**Review by Kegsay — COMMENTED** [review]
We need to implement the topological filtering bit.

> limit can be set both as a query parameter, and as a part of RoomEventFilter.
The spec doesn't seem to specify which takes precedence if both are specified.
For now I've assumed that the query parameter takes precedence.

Your assumption is the same as mine: `limit` takses precedence.

I would go for a 20 limit, not 10. 

> Unlike /sync, the spec doesn't seem to allow for the filter payload to be a filter_id
instead of a JSON RoomEventFilter. Is this intended, or a spec oversight?

Almost certainly a spec oversight. I didn't even know `?filter=` was A Thing on `/messages`!

**Comment by kegsay** [comment]
Stale. Please re-open when this has been brough up-to-date and the outstanding issues resolved. Thanks!

**Pull request closed without merging**